Support running the JLS with Java 9

We check the java version used to run the JLS is 1.8. We should ensure 1.9 works too.

And just when I thought this was over, I realized it fails with the latest JDK 9-ea release (I was testing with an old 1.9.0-ea-b63 build).
- the version prefix is different, that's easy to fix
- then I get an error similar to https://bugs.eclipse.org/bugs/show_bug.cgi?id=498728 on startup:

```
java.version=9-ea
java.vendor=Oracle Corporation
BootLoader constants: OS=macosx, ARCH=x86_64, WS=cocoa, NL=fr_FR
Framework arguments:  --add-modules=java.se.ee
Command-line arguments:  -data /Users/fbricon/Library/Application Support/Code - Insiders/User/workspaceStorage/b8821950c971f4d89c32232a9065b58c/redhat.java/jdt_ws --add-modules=java.se.ee

!ENTRY org.eclipse.core.resources 2 10035 2016-10-11 15:47:38.002
!MESSAGE The workspace exited with unsaved changes in the previous session; refreshing workspace to recover changes.

!ENTRY org.eclipse.osgi 4 0 2016-10-11 15:47:38.479
!MESSAGE Application error
!STACK 1
java.lang.NoClassDefFoundError: java/sql/Time
	at com.google.gson.Gson.<init>(Gson.java:240)
	at com.google.gson.GsonBuilder.create(GsonBuilder.java:569)
	at org.jboss.tools.langs.base.LSPServer.<init>(LSPServer.java:143)
	at org.jboss.tools.vscode.java.internal.JavaClientConnection.<init>(JavaClientConnection.java:71)
	at org.jboss.tools.vscode.java.internal.JavaLanguageServerPlugin.startConnection(JavaLanguageServerPlugin.java:49)
	at org.jboss.tools.vscode.java.internal.JavaLanguageServerPlugin.startLanguageServer(JavaLanguageServerPlugin.java:103)
	at org.jboss.tools.vscode.java.internal.LanguageServer.start(LanguageServer.java:26)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@9-ea/Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@9-ea/NativeMethodAccessorImpl.java:62)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@9-ea/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java.base@9-ea/Method.java:535)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:673)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:610)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1519)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1492)
Caused by: java.lang.ClassNotFoundException: java.sql.Time
	at java.lang.ClassLoader.findClass(java.base@9-ea/ClassLoader.java:665)
	at java.lang.ClassLoader.loadClass(java.base@9-ea/ClassLoader.java:491)
	at java.lang.ClassLoader.loadClass(java.base@9-ea/ClassLoader.java:424)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:360)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:353)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:161)
	at java.lang.ClassLoader.loadClass(java.base@9-ea/ClassLoader.java:424)
	... 20 more

```

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -41,6 +41,13 @@ function runJavaServer() : Thenable<StreamInfo> {
 				// suspend=y is the default. Use this form if you need to debug the server startup code:
 				//  params.push('-agentlib:jdwp=transport=dt_socket,server=y,address=1044');
 			}
+			if (requirements.java_version > 8) {
+				params.push('--add-modules=ALL-SYSTEM');
+				params.push('--add-opens');
+				params.push('java.base/java.util=ALL-UNNAMED');
+				params.push('--add-opens');
+				params.push('java.base/java.lang=ALL-UNNAMED');
+			}
 			params.push('-Declipse.application=org.eclipse.jdt.ls.core.id1');
 			params.push('-Dosgi.bundles.defaultStartLevel=4');
 			params.push('-Declipse.product=org.eclipse.jdt.ls.core.product');

--- a/src/requirements.ts
+++ b/src/requirements.ts
@@ -14,6 +14,7 @@ const JAVAC_FILENAME = 'javac' + (isWindows?'.exe':'');
 
 interface RequirementsData {
     java_home: string;
+    java_version: number;
 }
 
 interface ErrorData {
@@ -31,9 +32,9 @@ interface ErrorData {
  */
 export async function resolveRequirements(): Promise<RequirementsData> {
     let java_home = await checkJavaRuntime();
-    await checkJavaVersion(java_home);
+    let javaVersion = await checkJavaVersion(java_home);
     await checkServerInstalled();
-    return Promise.resolve({ 'java_home': java_home });
+    return Promise.resolve({ 'java_home': java_home, 'java_version': javaVersion});
 }
 
 function checkJavaRuntime(): Promise<any> {
@@ -81,11 +82,13 @@ function readJavaConfig() : string {
 function checkJavaVersion(java_home: string): Promise<any> {
     return new Promise((resolve, reject) => {
         cp.execFile(java_home + '/bin/java', ['-version'], {}, (error, stdout, stderr) => {
-            if (stderr.indexOf('1.8') < 0){
+            if (stderr.indexOf('version "9') > -1){
+                resolve(9);
+            } if (stderr.indexOf('1.8') < 0){
                 openJDKDownload(reject, 'Java 8 is required to run. Please download and install a JDK 8.');
             }
             else{
-                resolve(true);
+                resolve(8);
             }
         });
     });

